Convert excerpt to Preact (2/n)

robertknight opened the pull request:
**~~Depends on https://github.com/hypothesis/client/pull/1477~~**

Convert the `Excerpt` component to Preact. This is the control which wraps the annotation quote and body to limit their height, displaying expand/collapse controls if they exceed a threshold.

Rather than convert the existing implementation verbatim, this is a completely new implementation which should be easier to understand and use. Instead of requiring callers to provide an input property which represents the displayed data, which triggers a re-measurement of the content height if it changes, the new implementation observes the DOM directly for size changes.

This component renders caller-provided content (ie. it accepts a `children`
prop), which is not supported by the Preact <-> Angular bridge. Therefore it was
also necessary to create components (`AnnotationBody`, `AnnotationQuote`) that
encapsulate uses of `Excerpt` inside the `<annotation>` component.

 - #1477 adds a utility module to watch for changes in the size of a DOM node
 - Add new `Excerpt` implementation and remove the old one
 - Remove `excerpt-overflow-monitor` utility that is not used by the new
   implementation
 - Add `AnnotationBody` component to render an annotation's markup body
   inside a (new) excerpt and convert the Angular template for
   `<annotation>` (annotation.html) to use it.
 - Add `AnnotationQuote` component to render an annotation's quote
   inside an excerpt and convert `annotation.html` to use it

lyzadanger commented on src/sidebar/components/excerpt.js:
```diff
@@ -1,178 +1,189 @@
+
+  // Update the measured height of the content after the initial render and
+  // when the size of the content element changes.
+  const updateContentHeight = useCallback(() => {
```
This may just be because the logic is what it is (and as simplified as it already can be), but it took me a while to grok the whole picture between here (line 80) and about line 96—the flow of state, control and what triggers what.

Part of this is simply that the `useCallback` hook is somewhat weirdly-named, and I always have to remind myself that's it's basically just memo-izing a fn reference.

I'm not sure I have any _suggestions_ here, it may be as straightforward as it can be already!

robertknight commented on src/sidebar/components/excerpt.js:
```diff
@@ -1,178 +1,189 @@
+    setContentHeight(newContentHeight);
+
+    const isCollapsible =
+      newContentHeight > collapsedHeight + overflowHysteresis;
```
It turns out prettier removes them again if you do this, presumably because the parens do not appear in the AST. I agree about readability in this context, so I made use of [`// prettier-ignore`](https://prettier.io/docs/en/ignore.html).

robertknight commented on src/sidebar/components/excerpt.js:
```diff
@@ -1,178 +1,189 @@
+    onCollapsibleChanged({ collapsible: isCollapsible });
+  }, [collapsedHeight, onCollapsibleChanged, overflowHysteresis]);
+
+  useLayoutEffect(() => {
```
`useLayoutEffect` effects run before the element is painted, `useEffect` effects run afterwards. In this case we need to calculate the collapsed state before the user sees the component as otherwise they will see a flash of the component in the wrong state.

In general, any situation where a component needs to examine its DOM output (eg. to measure its size), and then possibly change the output as a result, should use `useLayoutEffect`.

See https://reactjs.org/docs/hooks-reference.html#timing-of-effects.

robertknight commented on src/sidebar/components/excerpt.js:
```diff
@@ -1,178 +1,189 @@
+
+  // Render the (possibly truncated) content and controls for
+  // expanding/collapsing the content.
+  const overflowing = contentHeight > collapsedHeight + overflowHysteresis;
```
No, just an inconsistency. I think my general practice is to add an `is` or `has` prefix if it is not "sufficiently obvious" from the rest of the name that it is a boolean. I'll add an `is` prefix for consistency.

robertknight commented on src/sidebar/components/excerpt.js:
```diff
@@ -1,178 +1,189 @@
+
+  return (
+    <div className="excerpt" style={contentStyle}>
+      <div test-name="excerpt-content" ref={contentElement}>
```
It has no special meaning. I just added it as a convenience to be able to find this element in tests, since the `<div>` has no other distinguishing attributes.

robertknight commented on src/sidebar/components/excerpt.js:
```diff
@@ -1,178 +1,189 @@
+  collapse = false,
+  collapsedHeight,
+  inlineControls = true,
+  onCollapsibleChanged = noop,
```
I think it depends on how the callback is used. If we only guard a call to it (eg. `if (fn) { fn(...) }`) this is quite useful. If on the other hand it triggers other logic, I wouldn't use it.

robertknight commented on src/sidebar/components/annotation.js:
```diff
@@ -562,10 +562,6 @@ function AnnotationController(
       return;
     }
     self.canCollapseBody = canCollapse;
-
```
The Angular <-> Preact wrapper ensures that any callbacks (ie. where a Preact component invokes a callback prop passed in by an Angular component) trigger a digest. See https://github.com/hypothesis/client/blob/b43ed1abf20b1b815a15b22124c084df8a9cb579/src/sidebar/util/wrap-react-component.js#L36

robertknight commented on src/sidebar/components/excerpt.js:
```diff
@@ -1,178 +1,189 @@
+ */
+function Excerpt({
+  children,
+  collapse = false,
```
I expanded the `collapse` prop docs to clarify that it is used to control the collapsed/expanded state if the Excerpt is configured not to provide internal controls for this.

robertknight commented:
Hi @lyzadanger. Thanks for the initial review. I have updated the PR accordingly.

> For my personal ergonomic comfort (if I were going to be the reviewer), one place to start might be to calve off just the observe-element-size util and its tests into a PR as that would give me a chance to dive into the nuts and bolts of that (and make sure my ResizeObserver and MutationObserver fu is up to snuff). But maybe the whole shebang is fine.

I'm happy to carve off that part if you'd prefer. Just give me a ping on Slack and let me know your preference.

lyzadanger commented:
@robertknight If you calve off that `observer` PR, I'll get it a speedy review.

robertknight commented:
Rebased. I could potentially split out the `Excerpt` implementation into a separate PR. Let me know if you want me to do that.

lyzadanger commented on src/sidebar/components/excerpt.js:
```diff
@@ -1,178 +1,193 @@
+  return (
+    <div className="excerpt__inline-controls">
+      <span className="excerpt__toggle-link">
+        <a
```
Wondering if a `<button>` element would make more sense here?

lyzadanger commented on src/sidebar/components/excerpt.js:
```diff
@@ -1,178 +1,193 @@
-  template: require('../templates/excerpt.html'),
+Excerpt.propTypes = {
+  /**
+   * The content to render inside the container.
```
I feel like we could probably get away with not commenting `children` props if we didn't feel like it. That said, I have no aversion to this comment.

robertknight replied:
I agree. I'll remove it.

lyzadanger commented on src/sidebar/components/test/annotation-body-test.js:
```diff
@@ -0,0 +1,33 @@
+    AnnotationBody.$imports.$restore();
+  });
+
+  it('displays the comment if `isEditing` is false', () => {
```
"the comment" meaning "the annotation body content?"

robertknight replied:
Yes.

lyzadanger commented on src/sidebar/components/test/excerpt-test.js:
```diff
@@ -1,209 +1,135 @@
+      >
+        {content}
+      </Excerpt>,
+      { attachTo: container }
```
Ooh, this `attachTo` `option` is handy.

lyzadanger commented on src/sidebar/index.js:
```diff
@@ -160,14 +167,6 @@ function startAngularApp(config) {
       'loggedOutMessage',
       wrapReactComponent(require('./components/logged-out-message'))
     )
-    .component(
```
Good job remembering to remove these.

robertknight commented on src/sidebar/components/annotation-quote.js:
```diff
@@ -0,0 +1,54 @@
+
+  /**
+   * The text that the annotation refers to. This is rendered as plain text
+   * (ie. HTML tags are rendered literally).
```
Yes, exactly.

robertknight commented on src/sidebar/components/annotation-body.js:
```diff
@@ -0,0 +1,96 @@
+  collapse: propTypes.bool,
+
+  /**
+   * If the
```
Looking at this again, the `hasContent` prop is badly named and possibly redundant. It controls whether the annotation body displays struck-through text or a "redacted text" placeholder (a series of grey bars) if the annotation has been moderated. I will add a comment for the moment and then address in a follow-up.

robertknight commented on src/sidebar/components/annotation-body.js:
```diff
@@ -0,0 +1,96 @@
+
+AnnotationBody.propTypes = {
+  /**
+   * Whether to limit the height of the annotation body if it is tall.
```
Yes, I'll update the doc.

robertknight commented on src/sidebar/components/annotation-body.js:
```diff
@@ -0,0 +1,96 @@
+
+  /**
+   * Callback invoked when the height of the rendered annotation body increases
+   * above or falls below the threshold at which the `collapse` prop will affect
```
I thought it useful to explicitly mention the `collapse` prop here in case someone wonders "what threshold"?

robertknight commented on src/sidebar/components/annotation-body.js:
```diff
@@ -0,0 +1,96 @@
+
+  /**
+   * Callback invoked when the user edits the content of the annotation body
+   * when `isEditing` is true.
```
Indeed. I'll remove the part of the sentence.

robertknight commented on src/sidebar/components/annotation-body.js:
```diff
@@ -0,0 +1,96 @@
+  onEditText: propTypes.func,
+
+  /**
+   * Callback invoked when the user clicks a space in a truncated annotation
```
I'll reword this to clarify.

robertknight commented on src/sidebar/components/annotation-body.js:
```diff
@@ -0,0 +1,96 @@
+          inlineControls={false}
+          onCollapsibleChanged={onCollapsibleChanged}
+          onToggleCollapsed={collapsed => onToggleCollapsed({ collapsed })}
+          overflowHystersis={20}
```
Gah. The term "hysteresis" was originally chosen as in "thermostat hysteresis". However, per Wikipedia, "Hysteresis is the dependence of the state of a system on its history". This prop doesn't really work like that though. It is just a fixed threshold above `collapsedHeight` before the excerpt is collapsed. I've renamed it to `overflowThreshold` instead.

In practice, it allows a ~4 line quote to be shown fully even though quotes are collapsed to 3 lines if they exceed that height. For annotation bodies, it is less important as the collapsed height is much taller.

robertknight commented on src/sidebar/components/annotation-quote.js:
```diff
@@ -0,0 +1,54 @@
+function AnnotationQuote({ isOrphan, quote, settings = {} }) {
+  return (
+    <section
+      className={classnames('annotation-quote-list', isOrphan && 'is-orphan')}
```
Looking at the API, an annotation can have multiple targets or multiple quote selectors within a target. AFAIK in Firefox it is, or was, technically possible to create a discontinuous text selection. We don't really support this though so it is obsolete.

robertknight commented on src/sidebar/components/annotation-body.js:
```diff
@@ -0,0 +1,96 @@
+          inlineControls={false}
+          onCollapsibleChanged={onCollapsibleChanged}
+          onToggleCollapsed={collapsed => onToggleCollapsed({ collapsed })}
+          overflowHystersis={20}
```
The value was chosen to roughly correspond to the height of a line in an annotation quote / body. If you were using it in a context with a larger font or a different type of content you might want a different value. IIRC the reason it isn't measured in units of line height is purely for reasons of implementation ease.

robertknight commented:
> Feel free to merge once the few actual action items (incomplete comments, Hysteresis typos) are addressed to your satisfaction.

Thanks Lyza. I've addressed the comments and will merge when tests complete. I made a note of some additional possible simplifications which I'll do as a follow-up.
